### PR TITLE
Mac: Enable daemon service to run as soon as user login

### PR DESCRIPTION
**Pull request by praveenkumar**
Without this PR following scenario happen
- User installed the crc first time and do the setup
- As part of the setup crc daemon is started by launchd
- User reboot
- Tray autostart with reboot but `crc setup --check-only` fails
because `launchd agent 'com.redhat.crc.daemon' is not running`.

With this PR, check for daemon task running should first validate
if an existing process of `crc daemon` is running and the version
it get from the api should be same as the binary then task running
check succeed. So this PR will makes `crc setup --check-only` work
again with tray autostart.

### CODE REVIEW

**Comment by cfergeau** [comment]
I thought launchd would ensure the daemon is started when the user opens a session on their mac? The launchd changes don't guarantee the daemon will be running during startup?

**Comment by praveenkumar** [comment]
> I thought launchd would ensure the daemon is started when the user opens a session on their mac? The launchd changes don't guarantee the daemon will be running during startup?

@cfergeau As of now launchd is working same way as we are doing in linux except the socket activation part (in linux our socket activation service is enabled but that doesn't start daemon until something query to that socket) here also I think we don't want to start the daemon by default by launchd, right now it is required because of tray but in future tray is not part of crc.

**Comment by cfergeau** [comment]
> @cfergeau As of now launchd is working same way as we are doing in linux except the socket activation part (in linux our socket activation service is enabled but that doesn't start daemon until something query to that socket) here also I think we don't want to start the daemon by default by launchd, right now it is required because of tray but in future tray is not part of crc.

On linux, the daemon is not automatically started on login because we can avoid it thanks to the socket activation files. If socket activation did not exist, we would have to start the daemon automatically on login.
If the daemon is not autostarted, I don't understand what launchd is used for? What does this bring over just running `crc daemon` manually?

**Comment by praveenkumar** [comment]
> If the daemon is not autostarted, I don't understand what launchd is used for? What does this bring over just running `crc daemon` manually.

Running daemon manually means you have to keep that terminal always open or manually run it in background, as user you have to make sure stop it when not required `crc cleanup` not remove that process also, so even with current state it is better than running manually but our plan is to have it auto start every restart that we can improve over time.

**Comment by cfergeau** [comment]
>  even with current state it is better than running manually but our plan is to have it auto start every restart that we can improve over time.

I am really unclear on what "the current state" is. I'm totally lost as to what happens if I run `crc setup`, and then reboot the machine. How will the daemon be started?

**Comment by praveenkumar** [comment]
> > even with current state it is better than running manually but our plan is to have it auto start every restart that we can improve over time.
> 
> I am really unclear on what "the current state" is. I'm totally lost as to what happens if I run `crc setup`, and then reboot the machine. How will the daemon be started? Isn't autostarting the daemon on session logging one or 2 lines to change in the launchd plist file?

@cfergeau I thought we don't want that but yes it is easy and I updated the PR accordingly.

File: pkg/os/launchd/launchd_darwin.go
Comment (anjannath) [inline pkg/os/launchd/launchd_darwin.go:117]: aha, does the `-w` mean its going to wait till the operation returns?

Comment (praveenkumar) [reply pkg/os/launchd/launchd_darwin.go:117]: No as per launchctl man page this suppose to enable the service 
```
              -w       Overrides the Disabled key and sets it to false or true for the load and unload subcommands respectively. In previous versions, this option would modify the
                       configuration file. Now the state of the Disabled key is stored elsewhere on- disk in a location that may not be directly manipulated by any process other than
                       launchd.

```

**Comment by cfergeau** [comment]
One limitation to the current approach is if the daemon was started by the tray in a previous version. Starting the daemon with launchd will silently fail, and nothing detects this old daemon version.
One way to solve this is to check the version of the running daemon as was done in an earlier iteration of this code, and kill any old version before restarting it in the proper version.

File: pkg/crc/preflight/preflight_checks_common.go
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_common.go:25]: `isDaemonRunningWithReleasedVersion` is almost the same code as `checkDaemonStarted` which is used on all platforms, so this could be factored.

File: pkg/crc/preflight/preflight_checks_common.go
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_common.go:166]: This function is called `isDaemonRunningWithReleasedVersion`, it's not supposed to stop/kill the daemon as a side effect.

File: pkg/crc/preflight/preflight_checks_common.go
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_common.go:202]: if any of the following conditions are met? if all of the following conditions are met? if none of the following conditions are met?

File: pkg/crc/preflight/preflight_checks_darwin.go
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_darwin.go:274]: What about writing it like this? This seems clearer regarding what the code wants to do.

```
if olderDaemonVersionRunning() {
   if err := killDaemonProcess(); err != nil {
       return err
   }
}
if !launchd.AgentRunning(daemonConfig.Label) {
			return fmt.Errorf("launchd agent '%s' is not running", daemonConfig.Label)
}
```

File: pkg/crc/preflight/preflight_checks_common.go
Comment (anjannath) [reply pkg/crc/preflight/preflight_checks_common.go:166]: The killing of the existing daemon process can be part of the fix function for this check?

File: pkg/crc/preflight/preflight_checks_common.go
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_checks_common.go:25]: @cfergeau Handled in #3099 , will hold this until that merge and then rebase it

File: pkg/crc/preflight/preflight_checks_darwin.go
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_checks_darwin.go:274]: This will be also easy as once #3099 in.

File: pkg/crc/preflight/preflight_checks_common.go
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_checks_common.go]: @cfergeau updated to `any of the following`.

File: pkg/crc/preflight/preflight_checks_darwin.go
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_checks_darwin.go:274]: @cfergeau It is now suggested way.

File: pkg/crc/preflight/preflight_checks_linux.go
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_checks_linux.go:362]: In linux I have to add it in this socket file check otherwise `cross-lint` error out not using `killDaemonProcess` function to linux. If we don't want to do that then might be we need to create a different file `preflight_check_nonlinux` and move this there.

**Comment by gbraad** [comment]
> [praveenkumar](https://github.com/praveenkumar) added the [has-to-be-in-release](https://github.com/code-ready/crc/labels/has-to-be-in-release) label [4 hours ago](https://github.com/code-ready/crc/pull/3095#event-6393606320)

Why was this tag added? Is everything lined up correctly to allow this?

**Comment by praveenkumar** [comment]
> > [praveenkumar](https://github.com/praveenkumar) added the [has-to-be-in-release](https://github.com/code-ready/crc/labels/has-to-be-in-release) label [4 hours ago](https://github.com/code-ready/crc/pull/3095#event-6393606320)
> 
> Why was this tag added? 

@gbraad because I want this to be part of 2.1.0 release (upcoming one).

> Is everything lined up correctly to allow this? 

Yes, tray side changes were already there, when I filed this PR initially I thought it would be a trival but @cfergeau suggestions around having launchd service as auto reload and also detecting the old version of daemon and kill it, took
some time to implement and getting it reviewed. 

> Which issue does this solve or related to?

This will solve the issue for a user who just want to do `crc setup` and `crc start` on mac with or without tray.

**Comment by gbraad** [comment]
I am just concerned about how this was planned. Feels a little last minute.

We have to better describe dependency to merge, or accept a known issue?

**Comment by anjannath** [comment]
> 

The review comments are mostly addressed, i'll test this now e2e with the tray and upgrade scenario

File: pkg/crc/preflight/preflight_checks_darwin.go
Comment (cfergeau) [reply pkg/crc/preflight/preflight_checks_darwin.go:274]: Anjan had a good point though that the killing fits better in a `fix` method rather than in `check`.

**Review by cfergeau — APPROVED** [review]
Can you add a description of the bug this fixes, or a link to the relevant issue? Apart  from this, looks good to me.

**Pull request merged**
